src/dashboard/apps/consent/helpers.py

```diff
@@ -76,3 +85,74 @@ def send_mail(recipients: list, template_id: int, email_data: dict) -> int | Non
         # fail silently and send a sentry log
         sentry_sdk.capture_exception(e)
         return 0
+
+
+def renew_expiring_consents():
+    """Renew expiring consents to ensure continuity.
+
+    Perform the renewal of expiring consents to ensure continuity for consents that
+    are nearing their expiration date and have not been renewed yet.
+    The function identifies consents that are set to expire shortly and generates
+    new consents starting immediately after the expiry date of the current consent,
+    with an extended validity period.
+
+    - Consents should be renewed if their end date is earlier than or equal to
+    `now + CONSENT_NUMBER_DAYS_END_DATE`.
+    - renewed consent date:
+        - start date of new consent is end date of current consent
+        - end date of new consent is calculate from CONSENT_NUMBER_DAYS_END_DATE
+
+    Returns:
+        List[Consent]: A list of newly created Consent objects that have been renewed
+         from the expiring consents.
+    """
+    # calculate the date from which to check
+    now = timezone.now()
+    end_of_today = now.replace(hour=23, minute=59, second=59, microsecond=999999)
+    checking_date = end_of_today + datetime.timedelta(days=CONSENT_UPCOMING_DAYS_LIMIT)
+
+    # calculate the end date of renewed consents
+    renewal_end_date = consent_end_date(CONSENT_NUMBER_DAYS_END_DATE)
+    if not CONSENT_NUMBER_DAYS_END_DATE:
+        renewal_end_date = renewal_end_date.replace(year=renewal_end_date.year + 1)
+
+    # Subquery to check if a renewed consent exists for validated consents.
+    # (this subquery only returns the `provider_assigned_id` of the consents found)
+    existing_renewed_consents = Consent.objects.filter(
+        start__gte=OuterRef("end"),
+        delivery_point=OuterRef("delivery_point"),
+        id_station_itinerance=OuterRef("id_station_itinerance"),
+        station_name=OuterRef("station_name"),
+        provider_assigned_id=OuterRef("provider_assigned_id"),
+    ).values("provider_assigned_id")
+
+    # Get expiring consents
+    #  = all validated consents - validated consents with existing consents.
+    # fmt: off
+    expiring_consents = (
+        Consent.validated_objects
+            .exclude(provider_assigned_id__in=Subquery(existing_renewed_consents))
+            .filter(end__lte=checking_date)
+    )
+    # fmt: on
```

Comment:
Let Black do its job :wink: 

```suggestion
    expiring_consents = (
        Consent.validated_objects
            .exclude(provider_assigned_id__in=Subquery(existing_renewed_consents))
            .filter(end__lte=checking_date)
    )
```